planning/behavior_path_lane_change_module/src/scene.cpp

```diff
@@ -842,226 +844,238 @@ PathWithLaneId NormalLaneChange::getPrepareSegment(
+    target_lane_objects, [&](const PredictedObject & object) {
+      return is_same_direction(object) && (is_within_vel_th(object) || is_ahead_of_ego(object));
+    });
+
+  utils::path_safety_checker::filterObjects(
+    other_lane_objects, [&](const PredictedObject & object) {
+      return is_within_vel_th(object) && is_same_direction(object) && is_ahead_of_ego(object);
+    });
+
+  utils::path_safety_checker::filterObjects(
+    current_lane_objects, [&](const PredictedObject & object) {
+      return is_within_vel_th(object) && is_same_direction(object) && is_ahead_of_ego(object);
+    });
+
+  LaneChangeLanesFilteredObjects lane_change_target_objects;
 
-  // objects in current lane
   const auto is_check_prepare_phase = check_prepare_phase();
-  for (const auto & obj_idx : target_obj_index.current_lane) {
-    const auto extended_object = utils::lane_change::transform(
-      objects.objects.at(obj_idx), common_parameters, *lane_change_parameters_,
-      is_check_prepare_phase);
-    target_objects.current_lane.push_back(extended_object);
-  }
+  std::for_each(target_lane_objects.begin(), target_lane_objects.end(), [&](const auto & object) {
+    auto extended_predicted_object = utils::lane_change::transform(
+      object, common_parameters, *lane_change_parameters_, is_check_prepare_phase);
+    lane_change_target_objects.target_lane.push_back(extended_predicted_object);
+  });
 
-  // objects in target lane
-  for (const auto & obj_idx : target_obj_index.target_lane) {
-    const auto extended_object = utils::lane_change::transform(
-      objects.objects.at(obj_idx), common_parameters, *lane_change_parameters_,
-      is_check_prepare_phase);
-    target_objects.target_lane.push_back(extended_object);
-  }
+  std::for_each(current_lane_objects.begin(), current_lane_objects.end(), [&](const auto & object) {
+    auto extended_predicted_object = utils::lane_change::transform(
+      object, common_parameters, *lane_change_parameters_, is_check_prepare_phase);
+    lane_change_target_objects.current_lane.push_back(extended_predicted_object);
+  });
 
-  // objects in other lane
-  for (const auto & obj_idx : target_obj_index.other_lane) {
-    const auto extended_object = utils::lane_change::transform(
-      objects.objects.at(obj_idx), common_parameters, *lane_change_parameters_,
-      is_check_prepare_phase);
-    target_objects.other_lane.push_back(extended_object);
-  }
+  std::for_each(other_lane_objects.begin(), other_lane_objects.end(), [&](const auto & object) {
+    auto extended_predicted_object = utils::lane_change::transform(
+      object, common_parameters, *lane_change_parameters_, is_check_prepare_phase);
+    lane_change_target_objects.other_lane.push_back(extended_predicted_object);
+  });
 
-  return target_objects;
+  lane_change_debug_.filtered_objects = lane_change_target_objects;
+
+  return lane_change_target_objects;
+}
+
+void NormalLaneChange::filterObjectsAheadTerminal(
```

Comment:
@zulfaqar-azmi-t4 
Just a minor comment, you can ignore.

I haven't been able to review the entire code due to big changes, and not so familiar with this module,
but at first glance, it appears that the implementation generates polygons for all specified class's objects  and performs calculations on every point of the polygon in the lane coordinate system. 
I think that filtering out clearly distant objects using the Euclidean distance in the base_link coordinate system before generating polygons and later calculating in the lane coordinate system could potentially reduce computational cost.

Reply:
Thank you for the suggestion. 
Possibly, I don't want to make changes to this function in this PR.
It is because the lanelet based filtering target is the main objective of this PR.
I  was hoping to move this function elsewhere and do a bit refactoring later.

Reply:
ok 
> Possibly, I don't want to make changes to this function in this PR.

That's not a problem at all for me